### PR TITLE
assets to s3

### PR DESCRIPTION
**Pull request by AndreaFrancis**
Moving assets to s3.
Second part of https://github.com/huggingface/datasets-server/issues/1406

### CODE REVIEW

File: libs/libcommon/src/libcommon/config.py [hunk 1]
```diff
@@ -42,12 +42,14 @@
 
 ASSETS_BASE_URL = "assets"
 ASSETS_STORAGE_DIRECTORY = None
+ASSETS_S3_FOLDER_NAME = "assets"
```
Comment (severo) [inline libs/libcommon/src/libcommon/config.py:45]: it's a bit weird to have two configuration variables (ASSETS_STORAGE_DIRECTORY and ASSETS_S3_FOLDER_NAME), while we will only use one of them.

But I understand that this will be fixed in another PR where we will abstract S3 or FS storages in a common interface.

File: chart/templates/worker/_container.tpl [hunk 2]
```diff
@@ -7,6 +7,7 @@
   imagePullPolicy: {{ .Values.images.pullPolicy }}
   env:
   {{ include "envAssets" . | nindent 2 }}
+  {{ include "envS3" . | nindent 2 }}
```
Comment (severo) [inline chart/templates/worker/_container.tpl:10]: no other containers need these environment variables? the jobs, the api/admin services, the storage machine?

File: services/worker/pyproject.toml [hunk 3]
```diff
@@ -58,6 +58,7 @@ flake8 = "^3.9.2"
 flake8-pep585 = "^0.1.7"
 flake8-unused-arguments = "^0.0.13"
 isort = "^5.12.0"
+moto = "^4.2.5"
```
Comment (severo) [inline services/worker/pyproject.toml:61]: (maybe) install in the dev dependencies of libcommon instead

**Review by severo — COMMENTED** [review]
Another comment: could you describe how we will migrate from NFS to S3?

File: chart/templates/worker/_container.tpl (hunk 2 above)
Comment (AndreaFrancis) [reply chart/templates/worker/_container.tpl:10]: S3Client is initialized only in `worker` `rows` and `search`. I think we don't need it in other containers.

File: libs/libcommon/src/libcommon/config.py (hunk 1 above)
Comment (AndreaFrancis) [reply libs/libcommon/src/libcommon/config.py:45]: Yes, I think using `fsspec` would help us create the file directly in a specified location and ASSETS_STORAGE_DIRECTORY will be removed.

File: chart/templates/worker/_container.tpl (hunk 2 above)
Comment (severo) [reply chart/templates/worker/_container.tpl:10]: OK, thanks for checking

File: services/worker/pyproject.toml (hunk 3 above)
Comment (severo) [reply services/worker/pyproject.toml:61]: (feel free to ignore)

**Comment by AndreaFrancis** [comment]
> Another comment: could you describe how we will migrate from NFS to S3?

In Prod:
- We will need to stop the workers while @rtrompier  helps us move/copy the files from NFS to S3
- Need to enable DNS for assets to S3
- Merge and deploy this PR
The last two steps can be done at the same time or independently.

**Comment by severo** [comment]
but the assets URLs in `splits-first-rows-...` responses will still be valid once migrated to S3?

File: services/worker/pyproject.toml (hunk 3 above)
Comment (AndreaFrancis) [reply services/worker/pyproject.toml:61]: I think I will leave it as it is now, anyway once we use `fsspec` instead we will probably remove all the references to boto3 and moto3.

**Comment by AndreaFrancis** [comment]
> but the assets URLs in splits-first-rows-... responses will still be valid once migrated to S3?

I see that the way we store the URLS is like https://datasets-server.huggingface.co/assets/asoria/image/--/default/train/0/image/image.jpg
So, same as in rows https://datasets-server.huggingface.co/cached-assets/asoria/image/--/default/train/0/image/image.jpg the switch in the DNS to point https://datasets-server.huggingface.co/assets to s3 /assets should work  correctly as well.
Can you confirm @rtrompier?

**Comment by severo** [comment]
> We will need to stop the workers while @rtrompier helps us move/copy the files from NFS to S3

is there a way to copy everything, then when we stop the workers, only do a last sync for the few files that have changed? It would reduce the downtime for the workers.

**Comment by rtrompier** [comment]
> > but the assets URLs in splits-first-rows-... responses will still be valid once migrated to S3?
> 
> I see that the way we store the URLS is like https://datasets-server.huggingface.co/assets/asoria/image/--/default/train/0/image/image.jpg So, same as in rows https://datasets-server.huggingface.co/cached-assets/asoria/image/--/default/train/0/image/image.jpg the switch in the DNS to point https://datasets-server.huggingface.co/assets to s3 /assets should work correctly as well. Can you confirm @rtrompier?

Yes I confirm. I will only redirect the traffic on `/assets/*` on S3.

> is there a way to copy everything, then when we stop the workers, only do a last sync for the few files that have changed? It would reduce the downtime for the workers.

It's what I'm trying to do. 
I will start a first transfert will all files, and once deployed, I will start an incremental transfert.
I come back to you when I found the best solution ;)

**Review by severo — APPROVED** [review]

**Comment by severo** [comment]
hmmmm are we ready to deploy this????????

**Comment by severo** [comment]
I mean: is the migration finished?

**Comment by AndreaFrancis** [comment]
> hmmmm are we ready to deploy this????????

Yes, we will be working on the deploy with @rtrompier

**Comment by severo** [comment]
how long do you think it will take? (because meanwhile we cannot deploy anything)

**Comment by rtrompier** [comment]
> how long do you think it will take? (because meanwhile we cannot deploy anything)

Should be quick. We just performing some tests on staging env before.
